### PR TITLE
Add getblockdeltas function on Pastel.

### PR DESCRIPTION
**Pull request by tsimafeisetchinava000**
- Add getblockdeltas function on Pastel source code.
   https://github.com/zcash/zcash/search?q=getblockdeltas

### CODE REVIEW

File: src/txmempool.cpp [hunk 1]
```diff
@@ -120,6 +120,16 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
     return true;
 }
 
+bool CTxMemPool::getSpentIndex(const CSpentIndexKey &key, CSpentIndexValue &value)
+{
+    LOCK(cs);
+    std::map<CSpentIndexKey, CSpentIndexValue, CSpentIndexKeyCompare>::iterator it = mapSpent.find(key);
```
Comment (akobrin1) [inline src/txmempool.cpp:126]: Definitely need to use auto for such long iterator expressions

**Review by akobrin1 — CHANGES_REQUESTED** [review]

File: src/script/script.h [hunk 2]
```diff
@@ -566,8 +568,18 @@ class CScript : public CScriptBase
      */
     unsigned int GetSigOpCount(const CScript& scriptSig) const;
 
+    // insightexplorer, there may be more script types in the future
```
Comment (akobrin1) [inline src/script/script.h:571]: use enum class ScriptType, and change in the code ScriptType::P2PKH, etc...

File: src/script/script.cpp [hunk 3]
```diff
@@ -260,3 +272,22 @@ std::string CScript::ToString() const
     }
     return str;
 }
+
+// insightexplorer
+uint160 CScript::AddressHash() const
```
Comment (akobrin1) [inline src/script/script.cpp:277]: looks like a candidate for "noexcept"

File: src/spentindex.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(txid);
+        READWRITE(inputIndex);
+        READWRITE(blockHeight);
+        READWRITE(satoshis);
+        READWRITE(addressType);
+        READWRITE(addressHash);
+    }
+
+    CSpentIndexValue(uint256 t, unsigned int i, int h, CAmount s, int type, uint160 a) {
```
Comment (akobrin1) [inline src/spentindex.h:59]: mark all parameters const

File: src/spentindex.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    void SetNull() {
+        txid.SetNull();
+        inputIndex = 0;
+        blockHeight = 0;
+        satoshis = 0;
+        addressType = 0;
+        addressHash.SetNull();
+    }
+
+    bool IsNull() const {
```
Comment (akobrin1) [inline src/spentindex.h:81]: noexcept

**Review by akobrin1 — CHANGES_REQUESTED** [review]

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -1026,6 +1027,178 @@ UniValue reconsiderblock(const UniValue& params, bool fHelp)
...
+                CTxDestination dest = DestFromAddressHash(spentInfo.addressType, spentInfo.addressHash);
+                if (IsValidDestination(dest)) {
+                    delta.pushKV("address", keyIO.EncodeDestination(dest));
+                }
+                delta.pushKV("satoshis", -1 * spentInfo.satoshis);
+                delta.pushKV("index", (int)j);
+                delta.pushKV("prevtxid", input.prevout.hash.GetHex());
+                delta.pushKV("prevout", (int)input.prevout.n);
+
+                inputs.push_back(delta);
```
[the remark this answers is not in the document]
Comment (tsimafeisetchinava000) [reply src/rpc/blockchain.cpp:1075]: `inputs.push_back(delta) => inputs.push_back(std::move(delta));`

File: src/script/script.cpp [hunk 7]
```diff
@@ -260,3 +272,22 @@ std::string CScript::ToString() const
...
+        start = 3;
+    else if (this->IsPayToScriptHash())
+        start = 2;
+    else {
+        // unknown script type; return zeros (this can happen)
+        vector<unsigned char> hashBytes;
+        hashBytes.resize(20);
+        return uint160(hashBytes);
+    }
+    vector<unsigned char> hashBytes(this->begin()+start, this->begin()+start+20);
```
[the remark this answers is not in the document]
Comment (tsimafeisetchinava000) [reply src/script/script.cpp:291]: src/prevector.h : line 278
```
iterator begin() { return iterator(item_ptr(0)); }
const_iterator begin() const { return const_iterator(item_ptr(0)); }
```

so I didn't replace begin() to cbegin()

**Comment by akobrin1** [comment]
Please rebase on current master branch. Note that serialization template methods were changed. 
Resolve conflicts in script.h & main.h.

**Review by a-ok123 — APPROVED** [review]

**Pull request closed without merging**